# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/cairo/cairo/src/cairo-tag-stack.c
# Commit: 3cdfb39376b7
# Full Hash: 3cdfb39376b736597fe658a322389117686b7e82
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2024-05-07 09:13:07
# Regressor Bug: 1892913
# File Overlap Count: 3
# Description:
#   Bug 1892913 - patch 1 - Update cairo source to 1.18.0 r=gfx-reviewers,lsalzman
#   
#   Wholesale replacement of files in gfx/cairo/cairo with contents of the 1.18.0 release tarball,
#   omitting subdirectories that are not part of the core library build.
#   
# ==============================================================================

diff -r 9daa351f875b -r 3cdfb39376b7 gfx/cairo/cairo/src/cairo-tag-stack.c
--- a/gfx/cairo/cairo/src/cairo-tag-stack.c	Mon May 06 19:22:35 2024 +0000
+++ b/gfx/cairo/cairo/src/cairo-tag-stack.c	Mon May 06 19:25:12 2024 +0000
@@ -99,6 +99,9 @@
     "Formula",
     "Form",
 
+    /* Section 14.8.2.2.2 - Artifacts */
+    "Artifact",
+
     NULL
 };
 
@@ -106,6 +109,8 @@
 static const char * _cairo_tag_stack_cairo_tag_list[] =
 {
     CAIRO_TAG_DEST,
+    CAIRO_TAG_CONTENT,
+    CAIRO_TAG_CONTENT_REF,
     NULL
 };
 
@@ -166,6 +171,17 @@
 	return _cairo_tag_error ("Invalid tag: %s", name);
     }
 
+    cairo_tag_stack_elem_t *top = _cairo_tag_stack_top_elem (stack);
+    if (top &&
+	(strcmp (top->name, CAIRO_TAG_CONTENT) == 0 ||
+	 strcmp (top->name, CAIRO_TAG_CONTENT_REF) == 0 ||
+	 strcmp (top->name, "Artifact") == 0))
+    {
+	return _cairo_tag_error ("%s tag can not contain nested tags",
+				 (strcmp (top->name, CAIRO_TAG_CONTENT) == 0) ? "CAIRO_TAG_CONTENT" :
+				 ((strcmp (top->name, CAIRO_TAG_CONTENT_REF) == 0) ? "CAIRO_TAG_CONTENT_REF" : top->name));
+    }
+
     if (stack->type == TAG_TREE_TYPE_NO_TAGS) {
 	if (name_in_list (name, _cairo_tag_stack_tagged_pdf_top_level_element_list))
 	    stack->type = TAG_TREE_TYPE_TAGGED;
@@ -259,6 +275,19 @@
 }
 
 void
+_cairo_tag_stack_foreach (cairo_tag_stack_t *stack,
+			  void (*func)(cairo_tag_stack_elem_t *elem,
+				       void *closure),
+			  void *closure)
+{
+    cairo_tag_stack_elem_t *elem;
+
+    cairo_list_foreach_entry (elem, cairo_tag_stack_elem_t, &stack->list, link) {
+	func (elem, closure);
+    }
+}
+
+void
 _cairo_tag_stack_free_elem (cairo_tag_stack_elem_t *elem)
 {
     free (elem->name);
@@ -276,9 +305,15 @@
     if (strcmp(name, "Link") == 0)
 	return (TAG_TYPE_LINK | TAG_TYPE_STRUCTURE);
 
-    if (strcmp(name, "cairo.dest") == 0)
+    if (strcmp(name, CAIRO_TAG_DEST) == 0)
 	return TAG_TYPE_DEST;
 
+    if (strcmp(name, CAIRO_TAG_CONTENT) == 0)
+	return TAG_TYPE_CONTENT;
+
+    if (strcmp(name, CAIRO_TAG_CONTENT_REF) == 0)
+	return TAG_TYPE_CONTENT_REF;
+
     return TAG_TYPE_STRUCTURE;
 }
 